brooklyn-server/camp/camp-brooklyn/src/test/java/org/apache/brooklyn/camp/brooklyn/DslAndRebindYamlTest.java

```diff
@@ -115,34 +125,207 @@ public T call() throws Exception {
+                "services:\n" +
+                "- type: org.apache.brooklyn.entity.group.DynamicCluster\n" +
+                "  id: test-cluster\n" +
+                "  initialSize: 1\n" +
+                "  memberSpec:\n" +
+                "    $brooklyn:entitySpec:\n" +
+                "      type: org.apache.brooklyn.core.test.entity.TestEntity\n" +
+                "      brooklyn.config:\n" +
+                "        test.confName: $brooklyn:component(\"test-cluster\").attributeWhenReady(\"sensor\")";
+
+        final Entity testEntity = createAndStartApplication(yaml);
+
+        DynamicCluster clusterEntity1 = (DynamicCluster) Iterables.getOnlyElement(testEntity.getApplication().getChildren());
+
+        TestEntity testEntity1 = null;
+        for (Entity entity : clusterEntity1.getChildren()) {
+            if (entity instanceof TestEntity) {
+                testEntity1 = (TestEntity) entity;
+                break;
+            }
+        }
+        Assert.assertNotNull(testEntity1, "TestEntity not found in DynamicCluster");
+
+        final TestEntity childTestEntity = testEntity1;
+
+        // Wait for the attribute to be ready in a new Task
+        Callable<String> configGetter = new Callable<String>() {
+            @Override
+            public String call() throws Exception {
+                String s = getConfigInTask(childTestEntity, TestEntity.CONF_NAME);
+                getLogger().info("getConfig {}={}", TestEntity.CONF_NAME, s);
+                return s;
+            }
+        };
+        ExecutorService executorService = Executors.newSingleThreadExecutor();
+        Future<String> stringFuture = executorService.submit(configGetter);
+
+        // Persist and rebind
+        Application app2 = rebind(testEntity.getApplication());
+
+        DynamicCluster clusterEntity2 = (DynamicCluster) Iterables.getOnlyElement(app2.getApplication().getChildren());
+
+        TestEntity testEntity2 = null;
+        for (Entity entity : clusterEntity2.getChildren()) {
+            if (entity instanceof TestEntity) {
+                testEntity2 = (TestEntity) entity;
+                break;
+            }
+        }
+        Assert.assertNotNull(testEntity2, "TestEntity not found in DynamicCluster");
+
+        Maybe<Object> maybe = testEntity2.config().getLocalRaw(TestEntity.CONF_NAME);
+
+        Assert.assertTrue(maybe.isPresentAndNonNull());
+        Assert.assertTrue(BrooklynDslDeferredSupplier.class.isInstance(maybe.get()));
+        BrooklynDslDeferredSupplier deferredSupplier = (BrooklynDslDeferredSupplier) maybe.get();
+        Assert.assertEquals(deferredSupplier.toString(), "$brooklyn:entity(\"test-cluster\").attributeWhenReady(\"sensor\")");
+
+        // Check that the Task is still waiting for attribute to be ready
+        Assert.assertFalse(stringFuture.isDone());
+
+        // Now set sensor value
+        ((EntityInternal) clusterEntity1).sensors().set(Sensors.newStringSensor("sensor"), "bar");
```

Comment:
This seems to be confusing modifying the entities from before rebind, but doing so after rebind. The call to `rebind()` is to simulate us having restarted the Brooklyn server. It's configurable whether we actually terminate the original management context. I guess this test is passing because the original management context was not terminated.

I'm looking at these tests in more detail. I'll make a few changes and submit a pull request against your branch @googlielmo (either this evening or tomorrow).